File: algorithm/combinatorics.basic/permutation.py
#!/usr/bin/env python

def permutations(head, tail=''):
  if len(head) == 0:print tail
  else:
    for i in range(len(head)):
      permutations(head[0:i] + head[i+1:], tail + head[i])

permutations("123")

File: ruby/permutation.rb
#!/usr/bin/env ruby
# encoding: utf-8

class Array
  def permutations
    return [self] if size < 2
    perm = []
    each {|e| (self - [e]).permutations.each { |p| perm << ([e] +p) } }
    perm
  end
end


p [*1..4].permutations
